content/en/blog/_posts/2024-09-20-kubeadm-v1beta4.md

```diff
@@ -0,0 +1,100 @@
+`caCertificateValidityPeriod`, `encryptionAlgorithm`, `dns.disabled` and `proxy.disabled`.
+- Support `extraEnvs` for all control plan components
+- `extraArgs` changed from a map to structured extra arguments for duplicates
+- Add a `timeouts` structure for init, join, upgrade and reset.
+
+For details, you can see the official document below:
+
+- Support custom environment variables in control plane components under `ClusterConfiguration`.
+Use `apiServer.extraEnvs`, `controllerManager.extraEnvs`, `scheduler.extraEnvs`, `etcd.local.extraEnvs`.
+- The ResetConfiguration API type is now supported in v1beta4. Users are able to reset a node by passing
+a `--config` file to "kubeadm reset".
+- `dryRun` mode is now configurable in InitConfiguration and JoinConfiguration.
+- Replace the existing string/string extra argument maps with structured extra arguments that support duplicates.
+ The change applies to `ClusterConfiguration` - `apiServer.extraArgs`, `controllerManager.extraArgs`,
+ `scheduler.extraArgs`, `etcd.local.extraArgs`. Also to `nodeRegistrationOptions.kubeletExtraArgs`.
+- Add `ClusterConfiguration.encryptionAlgorithm` that can be used to set the asymmetric encryption algorithm
+used for this cluster's keys and certificates. Can be one of "RSA-2048" (default), "RSA-3072", "RSA-4096" or "ECDSA-P256".
+- Add `ClusterConfiguration.dns.disabled` and `ClusterConfiguration.proxy.disabled` that can be used to disable
+the CoreDNS and kube-proxy addons during cluster initialization. Skipping the related addons phases,
+during cluster creation will set the same fields to `true`.
+- Add the `nodeRegistration.imagePullSerial` field in `InitConfiguration` and `JoinConfiguration`, which
+can be used to control if kubeadm pulls images serially or in parallel.
+- The UpgradeConfiguration kubeadm API is now supported in v1beta4 when passing `--config` to "kubeadm upgrade" subcommands.
+For upgrade subcommands, the usage of component configuration for kubelet and kube-proxy, as well as InitConfiguration and
+ClusterConfiguration, is deprecated and will be ignored when passing `--config`.
+- Add a `timeouts` structure to `InitConfiguration`, `JoinConfiguration`, `ResetConfiguration` and `UpgradeConfiguration`
+that can be used to configure various timeouts. The `ClusterConfiguration.timeoutForControlPlane` field is replaced
+by `timeouts.controlPlaneComponentHealthCheck`. The `JoinConfiguration.discovery.timeout` is replaced by
+`timeouts.discovery`.
+- Add a `certificateValidityPeriod` and `caCertificateValidityPeriod` fields to `ClusterConfiguration`. These fields
+can be used to control the validity period of certificates generated by kubeadm during sub-commands such as `init`,
+`join`, `upgrade` and `certs`. Default values continue to be 1 year for non-CA certificates and 10 years for CA
+certificates. Only non-CA certificates continue to be renewable by `kubeadm certs renew`.
```

Comment:
That last sentence confuses me. Which of these do we mean:
- there is no change to which certificates you can renew using `kubeadm certs renew`; as in previous minor releases, only leaf certificates can be renewed.
- changed the behavior of `kubeadm certs renew`. In kubeadm v1.31, you can only renew leaf certificates; if you need to renew or replace any CA certificate (either the trust anchor or an intermediate certificate), you must now find an alternative mechanism.

Let's help readers by rewording, so that only one meaning makes sense.

Reply:
The first one.

No behavior change here.